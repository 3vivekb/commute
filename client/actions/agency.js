import fetchAction from '@conveyal/woonerf/fetch'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add agency')
export const createAgency = (newAgency) => {
  newAgency.id = uuid.v4()
  newAgency.organizations = []
  return [
    addLocally(newAgency),
    push(`/agency/${newAgency.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete agency')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/agency/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAgency = (id) => [deleteLocally(id), push('/')]

export const loadAgencies = () => fetchAction({
  url: '/api/agency',
  next: (res, err) => {
    if (!err) {
      return setAllLocally(res)
    }
  }
})

export const updateAgency = (agency) => [
  setLocally(agency),
  push(`/agency/${agency.id}`)
]
/* const updateOnServer = (agency) =>
  serverAction({
    url: `/api/agency/${agency.id}`,
    options: {
      body: JSON.stringify(agency),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set agency')
export const setAllLocally = createAction('set agencies')
