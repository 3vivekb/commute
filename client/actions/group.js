import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

// site stuff
const addGroup = createAction('add group')
export const createGroup = (newGroup, organizationId) => {
  newGroup.id = uuid.v4()
  return [
    addGroup({ group: newGroup, organizationId }),
    push(`/organizations/${organizationId}/group/${newGroup.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete group')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/group/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteGroup = (id, organizationId) => [
  deleteLocally(id),
  push(`/organizations/${organizationId}`)
]
