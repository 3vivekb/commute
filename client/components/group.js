import React, {Component} from 'react'

export default class Group extends Component {
  static propTypes = {}

  render () {
    return <h1>Group</h1>
  }
}
