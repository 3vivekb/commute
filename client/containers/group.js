import {connect} from 'react-redux'

import Group from '../components/group'

function mapStateToProps (state) {
  return {}
}

export default connect(mapStateToProps)(Group)
