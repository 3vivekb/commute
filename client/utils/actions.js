import fetchAction from '@conveyal/woonerf/fetch'
import qs from 'qs'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'

/**
 * Make generic model actions that communicate with a rest server
 *
 * @param  {Object} cfg An object with the following parameters:
 * - commands: An object with the keys being commands and the corresponding cfg of each command
 * -- redirectionStrategy: The redirection strategy to use upon success.
 *  Can be one of following: 'toEntity', 'toParent', 'toHome', undefined (no redirection)
 *  Defaults are as follows:
 *    Collection GET, GET: null
 *    Collection POST, PUT: toEntity
 *    DELETE: toParent
 * - parentKey: Key in model of parent's id
 *   (must be provided if using toParent redirection strategy)
 * - parentName: Name of app route to parent view
 *   (must be provided if using toParent redirection strategy)
 * - pluralName: Plural name of the model
 * - singularName: Singular name of the model
 * @return {Object}     The actions
 */
const makeGenericModelActions = (cfg) => {
  const {commands, parentKey, parentName, pluralName, singularName} = cfg
  const actions = {}

  // make local set actions
  const addLocally = createAction(`add ${singularName}`)
  const deleteLocally = createAction(`delete ${singularName}`)
  const setLocally = createAction(`set ${singularName}`)
  const setAllLocally = createAction(`set ${pluralName}`)

  const baseEndpoint = `/api/${singularName}`

  const redirectionStrategies = {
    'toEntity': (entity) => {
      return push(`/${singularName}/${entity._id}`)
    },
    'toHome': () => {
      return push('/')
    },
    'toParent': (entity) => {
      return push(`/${parentName}/${entity[parentKey]}`)
    }
  }

  /**
   * Do a redirect if needed
   *
   * @param  {Object} cfg Paremters as follows:
   *   - actions: The actions to append to
   *   - endpointCfg: The endpoint config
   *   - defaultStrategy: The default redire strategy to use (default: undefined)
   *   - redirectArgs: Arguments to provide to the redirectionStrategy
   */
  const doRedirectIfNecessary = (cfg) => {
    const redirectionStrategy = (cfg.endpointCfg.redirectionStrategy !== undefined
      ? cfg.endpointCfg.redirectionStrategy
      : cfg.defaultStrategy)

    if (redirectionStrategy) {
      const redirect = redirectionStrategies[redirectionStrategy](cfg.redirectArgs)
      if (redirect) {
        cfg.actions.push(redirect)
      }
    }
  }

  if (commands['Collection GET']) {
    actions.loadAll = (queryParams) => {
      // only include filteredKeys in querystring
      let queryString = ''
      if (queryParams) {
        queryString = qs.stringify(queryParams)
        if (queryString.length !== 0) {
          queryString = '?' + queryString
        }
      }

      // make request
      return fetchAction({
        next: (res, err) => {
          if (!err) {
            return setAllLocally(res.value)
          }
          // TODO handle error
        },
        url: baseEndpoint + queryString
      })
    }
  }

  if (commands['Collection POST']) {
    const endpointCfg = commands['Collection POST']
    actions.create = (newEntity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const createdEntity = res.value
            const actions = [
              addLocally(createdEntity)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toEntity',
              redirectArgs: createdEntity
            })
            return actions
          }
        },
        options: {
          body: newEntity,
          method: 'POST'
        },
        url: baseEndpoint
      })
    }
  }

  if (commands['DELETE']) {
    const endpointCfg = commands['DELETE']
    actions.delete = (entity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const actions = [
              deleteLocally(entity._id)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toParent',
              redirectArgs: entity
            })
            return actions
          }
        },
        options: {
          method: 'DELETE'
        },
        url: `${baseEndpoint}/${entity._id}`
      })
    }
  }

  if (commands['GET']) {
    actions.loadOne = (id) => fetchAction({
      next: (res, err) => {
        if (!err) {
          return setLocally(res.value)
        }
        // TODO handle error
      },
      url: `${baseEndpoint}/${id}`
    })
  }

  if (commands['PUT']) {
    const endpointCfg = commands['PUT']
    actions.update = (entity) => fetchAction({
      next: (res, err) => {
        if (err) {
          // TODO handle error
        } else {
          const updatedEntity = res.value
          const actions = [
            setLocally(updatedEntity)
          ]
          doRedirectIfNecessary({
            actions,
            endpointCfg,
            defaultStrategy: 'toEntity',
            redirectArgs: updatedEntity
          })
          return actions
        }
      },
      options: {
        body: entity,
        method: 'PUT'
      },
      url: `${baseEndpoint}/${entity._id}`
    })
  }

  return actions
}

export default makeGenericModelActions
