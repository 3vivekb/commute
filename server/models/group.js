const {Schema} = require('mongoose')

const Commuter = require('./commuter')

const groupSchema = new Schema({
  name: {
    required: true,
    type: String
  },
  organizationId: {
    ref: 'Organization',
    required: true,
    type: Schema.Types.ObjectId
  }
})

groupSchema.pre('remove', function (next) {
  // CASCADE DELETE
  Commuter.remove({ groupId: this._id }).exec()
  next()
})

module.exports = groupSchema
