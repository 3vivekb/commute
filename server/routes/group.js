const pick = require('lodash.pick')

const models = require('../models')
const makeRestEndpoints = require('./').makeRestEndpoints

module.exports = function makeRoutes (app) {
  makeRestEndpoints(app,
    {
      childModels: [{
        foreignKey: 'groupId',
        key: 'commuters',
        model: models.Commuter
      }],
      commands: {
        'Collection GET': {},
        'GET': {},
        'DELETE': {},
        'PUT': {}
      },
      model: models.Group,
      name: 'group'
    }
  )

  // rest-ish endpoints
  app.post('/api/group', (req, res) => {
    res.set('Content-Type', 'application/json')
    const handleErr = (err) => res.status(500).json({error: err})
    let body = req.body
    if (!Array.isArray(body)) {
      body = [body]
    }
    const groupsToCreate = body.map((newGroup) => pick(newGroup, ['name', 'organizationId']))
    models.Group.create(groupsToCreate)
      .then((groups) => {
        const createGroupCommutersPromises = groups.map((group, idx) => {
          return new Promise((resolve, reject) => {
            if (body[idx].commuters && body[idx].commuters.length > 0) {
              // insert commuters one-by-one to trigger pre-save hook
              Promise.all(body[idx].commuters.map((commuter) => {
                return models.Commuter.create(Object.assign(commuter, { groupId: group._id }))
              }))
                .then((data) => {
                  const output = Object.assign({ commuters: data.map((commuter) => commuter._id) }, group._doc)
                  resolve(output)
                })
                .catch(reject)
            } else {
              resolve(group)
            }
          })
        })
        Promise.all(createGroupCommutersPromises)
          .then((data) => {
            res.json(data)
          })
          .catch((err) => {
            handleErr(err)
          })
      })
      .catch(handleErr)
  })
}
