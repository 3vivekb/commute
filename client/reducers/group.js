import update from 'react-addons-update'

import {addToEntityMap, entityArrayToEntityIdArray} from '../utils/entities'
import {makeChildrenHandlers, makeGenericReducerHandlers} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['delete', 'set', 'set many'],
  name: {
    singular: 'group',
    plural: 'groups'
  }
})

reducers['add group'] = function (state, action) {
  const newGroup = {...action.payload}
  newGroup.commuters = newGroup.commuters.map((commuter) => commuter._id)
  return addToEntityMap(state, newGroup)
}

reducers['append commuters'] = function (state, action) {
  const affectedGroup = state[action.payload.groupId]
  const modifiedGroup = update(affectedGroup, {
    commuters: {
      $push: entityArrayToEntityIdArray(action.payload.commuters)
    }
  })
  return addToEntityMap(state, modifiedGroup)
}

Object.assign(reducers, makeChildrenHandlers({
  childPluralName: 'commuters',
  childSingularName: 'commuter',
  parentIdField: 'groupId'
}))

export const initialState = {}
