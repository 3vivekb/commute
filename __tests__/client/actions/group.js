/* global describe, it */

import {expectCreateAction, expectDeleteGroup, expectUpdateGroup} from '../../test-utils/actions'

import * as group from '../../../client/actions/group'

describe('actions > group', () => {
  it('create group should work', () => {
    const data = {
      name: 'Mock Group'
    }
    const result = group.createGroup(data, 'organization-id')

    expectCreateAction(result)
  })

  it('delete group should work', () => {
    const result = group.deleteGroup('group-id', 'organization-id')

    expectDeleteGroup(result)
  })

  it('update group should work', () => {
    const data = {
      id: 'group-id',
      name: 'New Name'
    }
    const actions = group.updateGroup(data)

    expectUpdateGroup(actions)
  })
})
