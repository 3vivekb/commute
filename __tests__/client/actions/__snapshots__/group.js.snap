exports[`actions > group create group should work 1`] = `
Object {
  "payload": Object {
    "allAddressesGeocoded": false,
    "name": "Mock Group",
  },
  "type": "add group",
}
`;

exports[`actions > group create group should work 2`] = `
Object {
  "payload": Object {
    "args": Array [
      "/group/new-entity-id",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;

exports[`actions > group delete group should work 1`] = `
Array [
  Object {
    "payload": "group-id",
    "type": "delete group",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/organizations/organization-id",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;

exports[`actions > group update group should work 1`] = `
Object {
  "payload": Object {
    "id": "group-id",
    "name": "New Name",
  },
  "type": "update group",
}
`;
